### DISCUSSION TITLE
Validating expression without executing it

### DISCUSSION THREAD

**Discussion by dkekesi** [bijington/expressive#112] [Q&A]
What is the proper way to syntactically validate an expression without executing it? Or just execute it and catch exceptions?
Is there an error message returned from the validation method that aids in fixing errors (if any) or do I have to construct my own message?

**Comment by bijington** [comment]
> What is the proper way to syntactically validate an expression without executing it? Or just execute it and catch exceptions?
> Is there an error message returned from the validation method that aids in fixing errors (if any) or do I have to construct my own message?

@dkekesi currently the only way to validate is by calling `Evaluate` or `ReferencedVariables` and catch the exception that gets thrown. All exceptions are wrapped in an `ExpressiveException` and it should either provide a helpful message or the `InnerException` should provide more detail.

On the topic of the helpful message, most errors encountered when compiling the expression should be helpful (e.g. pointing out missing characters, parameters, etc.). There are some occasions when the actual evaluation can provide something less helpful as there are parts that currently just pass up the .NET exception details. If you do encounter some of these please do feel free to raise them and we can see if we can improve on the handling of those parts.

I would like to get to a point where there could be a method of validating without having to rely on Exception handling so I will probably proceed to raising an issue to cover this.

**Reply by dkekesi** [reply]
Thanks for the help. NCalc has an Expression.HasErrors() function making the check easy, although it provides no error messages. I guess your method is better as it will return messages, too (although in English only, I presume).

**Reply by bijington** [reply]
Oh great thanks for the pointer. I think some level of combining the exception handling messages with NCalcs approach could be the best way to tackle this.

Yes currently the framework only reports messages in English, do you have a potential use case to need other languages?

**Reply by dkekesi** [reply]
We need to provide errors in Hungarian and English, but in case of NCalc it was easy as the Expression.HasErrors() function returns Bool and thus the displayed message could be easily formulated. We will make a similar function with Expressive that will catch the exceptions upon parsing the expressions.

**Reply by bijington** [reply]
> We need to provide errors in Hungarian and English, but in case of NCalc it was easy as the Expression.HasErrors() function returns Bool and thus the displayed message could be easily formulated. We will make a similar function with Expressive that will catch the exceptions upon parsing the expressions.

You could wrap up the exception handling inside an extension method then it would appear very similar to NCalc for now. An alternative would be to provide a message based on the type of exception that you get back. There are a number of known Expressive exceptions and some will give a little bit of context to the issue e.g. FunctionNameAlreayRegisteredException
